pkg/agent/nodeportlocal/k8s/controller.go

```diff
@@ -112,12 +167,109 @@ func (c *Controller) EnqueueObj(obj interface{}) {
+	deletedState, ok := obj.(cache.DeletedFinalStateUnknown)
+	if !ok {
+		return nil, fmt.Errorf("Received unexpected object: %v", obj)
+	}
+	svc, ok := deletedState.Obj.(*corev1.Service)
+	if !ok {
+		return nil, fmt.Errorf("DeletedFinalStateUnknown object is not of type Service: %v", deletedState.Obj)
+	}
+	return svc, nil
+}
+
+func (c *NPLController) enqueueSvcUpdate(oldObj, newObj interface{}) {
+	// In case where the app selector in Service gets updated from one valid selector to another
+	// both sets of Pods (corresponding to old and new selector) need to be considered
+
+	// Donot push to queue if 1) Service ResourceVersions don't change OR
+	// 2) Service spec selectors AND 3) NPLEnabledAnnotationKey values don't change
+	newSvc := newObj.(*corev1.Service)
+	oldSvc := oldObj.(*corev1.Service)
+	if oldSvc.ResourceVersion == newSvc.ResourceVersion ||
+		reflect.DeepEqual(oldSvc.Spec.Selector, newSvc.Spec.Selector) &&
+			oldSvc.Annotations[NPLEnabledAnnotationKey] == newSvc.Annotations[NPLEnabledAnnotationKey] {
+		return
+	}
+
+	podKeys := append(c.getPodsFromService(oldSvc), c.getPodsFromService(newSvc)...)
+	for _, podKey := range podKeys {
+		c.queue.Add(podKey)
+	}
+}
+
+func (c *NPLController) enqueueSvc(obj interface{}) {
+	svc, isSvc := obj.(*corev1.Service)
+	if !isSvc {
+		var err error
+		svc, err = c.checkDeletedSvc(obj)
+		if err != nil {
+			klog.Errorf("Got error while processing event update: %v", err)
+			return
+		}
+	}
+
+	for _, podKey := range c.getPodsFromService(svc) {
+		c.queue.Add(podKey)
+	}
+}
+
+func (c *NPLController) getPodsFromService(svc *corev1.Service) []string {
+	var pods []string
+	annotations := svc.GetAnnotations()
+	if val, ok := annotations[NPLEnabledAnnotationKey]; !ok || val != "true" {
+		return pods
+	}
+
+	podList, err := c.podLister.List(labels.SelectorFromSet(labels.Set(svc.Spec.Selector)))
+	if err != nil {
+		klog.Errorf("Got error while listing Pods: %v", err)
+		return pods
+	}
+	for _, pod := range podList {
+		pods = append(pods, podKeyFunc(pod))
+	}
+	return pods
+}
+
+func (c *NPLController) isNPLEnabledForServiceOfPod(obj interface{}) bool {
+	pod := obj.(*corev1.Pod)
+	services, err := c.svcInformer.Informer().GetIndexer().ByIndex(NPLEnabledAnnotationIndex, "true")
+	if err != nil {
+		klog.Errorf("Got error while listing Services with annotation: %v", err)
+		return false
+	}
+
+	for _, service := range services {
+		svc, isSvc := service.(*corev1.Service)
+		// selecting Services NOT of type NodePort, with Service selector matching Pod labels
+		if isSvc && svc.Spec.Type != corev1.ServiceTypeNodePort {
+			if matchSvcSelectorPodLabels(svc.Spec.Selector, pod.GetLabels()) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// matchSvcSelectorPodLabels verifies that all key/value pairs present in Service's selector
+// are also present in Pod's labels
+func matchSvcSelectorPodLabels(svcSelector, podLabel map[string]string) bool {
```

Comment:
If selector has no selector, it is assumed to have an external process managing its endpoints, and should not select any Pod. 
See https://kubernetes.io/docs/concepts/services-networking/service/#without-selectors